slf4j/src/main/java/com/google/common/flogger/backend/slf4j/Slf4jLoggerBackend.java

```diff
@@ -0,0 +1,188 @@
+/*
+ * Copyright (C) 2018 The Flogger Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.common.flogger.backend.slf4j;
+
+import com.google.common.flogger.backend.LogData;
+import com.google.common.flogger.backend.LoggerBackend;
+import com.google.common.flogger.backend.SimpleMessageFormatter;
+import org.slf4j.Logger;
+
+import java.util.logging.Level;
+
+/**
+ * A logging backend that uses slf4j to output log statements.
+ */
+final class Slf4jLoggerBackend extends LoggerBackend implements
+    SimpleMessageFormatter.SimpleLogHandler {
+
+  private final Logger logger;
+
+  Slf4jLoggerBackend(Logger logger) {
+    if (logger == null) {
+      throw new NullPointerException("logger is required");
+    }
+    this.logger = logger;
+  }
+
+  // represents the log levels supported by SLF4J, used internally to dispatch calls accordingly
+  private enum Slf4jLogLevel {
+    TRACE, DEBUG, INFO, WARN, ERROR
+  }
+
+  /**
+   * Adapts the JUL level to SLF4J level per the below mapping:
+   *
+   * <table>
+   * <tr>
+   * <th>JUL</th>
+   * <th>SLF4J</th>
+   * </tr>
+   * <tr>
+   * <td>FINEST</td><td>TRACE</td>
+   * </tr><tr>
+   * <td>FINER</td><td>TRACE</td>
+   * </tr>
+   * <tr>
+   * <td>FINE</td><td>DEBUG</td>
+   * </tr>
+   * <tr>
+   * <td>CONFIG</td><td>DEBUG</td>
+   * </tr>
+   * <tr>
+   * <td>INFO</td>
+   * <td>INFO</td>
+   * </tr>
+   * <tr>
+   * <td>WARNING</td>
+   * <td>WARN</td>
+   * </tr>
+   * <tr>
+   * <td>SEVERE</td>
+   * <td>ERROR</td>
+   * </tr>
+   * </table>
+   *
+   * <p>Custom JUL levels are mapped to the next-lowest standard JUL level; for example, a custom
+   * level at 750 (between INFO:800 and CONFIG:700) would map to the same as CONFIG (DEBUG).</p>
+   *
+   * <p>It isn't expected that the JUL levels 'ALL' or 'OFF' are passed into this method;
```

Comment:
I double checked the Flogger code, and I don't actually have a check for this. We fixed callers inside Google with a static code check, but the library itself doesn't complain. In general there's a "no exception" policy for Flogger, since you don't want the act of turning log statements on to trigger a crash (since the "bad" code would not be run until logging was on).

However in this case I'd be okay with an exception, since the level must be passed and would be checked even when logging is off (the isLoggable() method is very "hot" and will get called for absolutely every log statement, so avoiding work there is essential).

I just wanted to clarify since you _might_ hit people passing ALL and OFF into Flogger. I will look at adding the check in the core library so everyone gets the same behaviour regardless of backends.